src/libsyntax_pos/hygiene.rs

```diff
@@ -337,7 +337,39 @@ pub enum ExpnFormat {
     /// e.g. `format!()`
     MacroBang(Symbol),
     /// Desugaring done by the compiler during HIR lowering.
-    CompilerDesugaring(Symbol)
+    CompilerDesugaring(CompilerDesugaringKind)
+}
+
+/// The kind of compiler desugaring.
+#[derive(Clone, Hash, Debug, PartialEq, Eq)]
+pub enum CompilerDesugaringKind {
+    BackArrow,
+    DotFill,
+    QuestionMark,
+}
+
+impl CompilerDesugaringKind {
+    pub fn as_symbol(&self) -> Symbol {
+        use CompilerDesugaringKind::*;
+        let s = match *self {
+            BackArrow => "<-",
+            DotFill => "...",
+            QuestionMark => "?",
+        };
+        Symbol::intern(s)
+    }
+}
+
+impl<'a> From<&'a str> for CompilerDesugaringKind {
```

Comment:
Somewhat along the lines of what @oli-obk was suggesting, I didn't expect to have `CompiledDesugaringKind::from`, but rather replace all places in the compiler where we use constant strings with direct references to the enum variants, and instead invoke `as_symbol()` anywhere that we need to print things out.